utils/vpr_io_place.py

```diff
@@ -24,6 +24,27 @@ def __init__(self):
         self.net_to_block = None
         self.net_map = {}
         self.inout_nets = set()
+        self.net_to_pad = set()
+
+    def read_io_loc_pairs(self, blif):
+        """
+         Read IO_LOC_PAIRS parameters from eblif carrying the information
+         which package pin a specified top port is constrained, e.g. O_LOC_PAIRS = "portA:D1"
+         In case of differential inputs/outputs there are two pairs of the parameter,
+         i.e. IO_LOC_PAIRS = "portA_p:D2,portA_n:D4"
+        """
+        if 'subckt' not in blif:
+            return
+        for attr in blif['subckt']:
+            if 'param' not in attr:
+                continue
+            if 'IO_LOC_PAIRS' in attr['param']:
+                locs = attr['param']['IO_LOC_PAIRS'][1:-1].split(',')
+                if 'NONE' in locs:
```

Comment:
Shouldn't this be part of the `for` loop

Reply:
Hmm, not necessarily. 'NONE' is added during the techmap step, so there won't be a `IO_LOC_PAIRS="NONE,NONE"` parameter. I only expect 'NONE' in `locs`.

Reply:
So you'd never expect `sig:A1,NONE`?   Only `NONE`?  In that case, it should be `'NONE' == locs` right?

Reply:
Correct. Otherwise someone might actually name his net NONE.

Reply:
I meant have NONE in the name

Reply:
I had another cold look at the code and in my opinion it's good as is. Since locs is a list as a result of the split the current solution is fine, i.e. `'NONE' in locs` matches only `NONE` so there is no risk of hitting a port name with `NONE` substring. Moreover `'NONE' == locs` won't work with list as it will always return false.
Rebased on master, so CI should turn green.

Reply:
Right.  In the event that IO_LOC_PAIRS="NONE", should we also check ` and len(locs) == 1`?

Reply:
I think `'NONE' in locs` is enough. The rest of the flow assumes it won't get any `NONE` values so adding `len(locs) == 1` opens the option to avoid this check with some manual eblif edit of the param to an illegal `port:D1,NONE`.